lib/spack/spack/test/concretize.py

```diff
@@ -152,7 +152,7 @@ class Root(Package):
     version("1.0", sha256="abcde")
     depends_on("changing")
 
-    conflicts("changing~foo")
+    conflicts("^changing~foo")
```

Comment:
> Is it not possible to declare a conflict on an arbitrary package anymore?

Declaring a conflict on an arbitrary package is already ill-posed in `develop`, since we treat:
```python
conflicts("foo ^bar")
```
the same way as:
```python
conflicts("^foo ^bar")
```
while they're clearly not the same. Here, both to clarify the semantic and to improve speed, I took the approach that `conflicts` can be declared only on direct dependencies, and only using a `^` before the dependency. The impact on the built-in repository is small (see #38748), so I would expect the impact on custom repositories to be equally small[^1]

This is in line with the local nature of all the other directives, and allows to compute the `condition_set` atoms faster. See also this comment in the code:

https://github.com/spack/spack/blob/89e7aa990b09df631ed755aeaf676a11f803ac62/lib/spack/spack/solver/concretize.lp#L268-L277

and #38748 which are related. In the latter PR, the only use case where it is sensible to have conflicts on an arbitrary node, is to model a package that vendors another. In that case, I think we should just add another directive:
```python
vendors("foo")
```
and encode separate rules for that.

**TL:DR** What I would do, is to add `vendor` directive and implement an audit that could flag the cases where we write:
```python
conflicts("foo")
```
but we mean:
```python
conflicts("^foo")
```

[^1]: Are we aware of repositories where people have a lot of conflicts on arbitrary packages in a recipe?